src/scalar/golden_ratio.rs

```diff
@@ -0,0 +1,145 @@
+//! Golden ratio is to minimization what bisection is to root finding.
+//! GoldenRatio searches within an interval for a local minimum. At 
+//! every iteration, the interval is decreased in size by a constant 
+//! factor, until the desired precision is obtained.
+//! 
+//! This algorithm is guaranteed to converge on a local minimum in a 
+//! finite amount of steps under very light smoothess criteria for the
+//! target function.
+//! 
+//! In an iteration, the target function is calculated at 4 points:
+//!         +---------+----+---------+
+//! iter 1  a         b    c         d
+//! The interval for the next iteration is chosen to be [a,c] if b<c,
+//! and [b,d] if c<=b. The distances b-a, c-b, and d-c are chosen in such
+//! a way that 3 out of 4 points can be reused, and only 1 new function 
+//! evaluation is required in the next iteration. If b<c this looks like: 
+//!         +---------+----+---------+
+//! iter 1  a         b    c         d
+//!         +----+----+----+
+//! iter 2  a    b    c    d
+
+#[derive(Builder, Debug)]
+pub struct GoldenRatio {
+    /// The width of the interval at which convergence is satisfactory. 
+    /// Smaller is more precise.
+    #[builder(default = "1e-8")]
+    pub xtol: f64,
+
+    /// The maximum number of iterations before the search terminates.
+    /// The number of function evaluations is 2+<number of iterations>.
+    /// Bigger is more precise.
+    #[builder(default = "1000")]
+    pub max_iter: usize,
+}
+
+/// The golden ratio is 1.0 : 0.5 + .5*5f64.sqrt(). Therefore we need the ratio
+/// 1.0 / (1.5 + .5*5f64.sqrt()) to calculate the fractions in which to divide
+/// intervals.
+const RATIO: f64 = 2.618033988749895; // 1.5 + 0.5*5f64.sqrt();
+
+// TODO improve the 'search around x0' to waste less function evaluations
+impl GoldenRatio {
+    
+    /// Search for the minimum of `func` around `x0`.
+    /// The search region is expanded in both directions until an expansion
+    /// leads to an increase in the function value at the bound of the region.
+    pub fn minimize<F>(&self, func: F, x0: f64) -> f64 
+    where F: Fn(f64) -> f64 {
+        let left = x0 + self.explore(&func, x0, -1.0);
```

Comment:
This is a tricky point. If you know your function has a sharp minimum close to x0 you can do a tight bracketed search. If you don't know this is the case, then you also wouldn't think of setting a small scale for these magic numbers. 

More generally, if you want to find a specific extreme point, the bracketed search should be your weapon of choice, while the x0-search only intends to find _some_ minimum. And if your function monotonically decreases outside [x0-epsilon, x0+epsilon] then I would argue that a minimizer routine should return `+/-inf`.

Reply:
I agree with returning `+/-inf` in such cases, but none the less there should not be such magic numbers in the code. I don't think it will pollute the builder too much, and also allows following good design principles

Reply:
I just realised that any constant is at risk of precision loss for large values of x0, rendering the result of `explore` meaningless.
How do you feel about using an ULPS as 'magic number'? 
* It takes only a couple of iterations extra with respect to the 1.0 magic number, 
* Is the most conditioned choice in the sense that a constant should actually be adapted to the size of x0, to account for the varying floating-point precision.

I'll make a commit to illustrate in the meantime.

Reply:
I decided to bound the ULPS from below by `f64::EPSILON`. since precision beyond ~16 decimals seems such a special case that the user should reconsider how (s)he defined his/her function.